### PR TITLE
Implement (contextual) keywords and use their versioning from v2

### LINKED ISSUE #568: Allow keywords to overlap identifiers
This is valid code in 0.4.21:

```solidity
contract ClientReceipt {
    event Deposit();
    function deposit() public payable {
        uint256 emit;
        emit Deposit();
    }
}
```

The key thing is that the emit statement is enabled from 0.4.21, but emit is still valid as an identifier. We currently cannot handle this, and it makes it impossible to compute over the versioning of tokens vs. identifiers.

### PR DESCRIPTION
**Pull request by Xanewok**
Closes #568 

There is still one outstanding issue where we return a `Vec<TokenKind>` from `next_token`; it'd like to return a more specialized type and ideally pass it on stack (2x2 bytes), rather than on-heap (extra 3x8 bytes for the Vec handle + indirection). We should name it better and properly show that we can return at most 2 token kinds (single token kind or identifier + kw combo).

To do:
- [x] Return tokens from `next_token` via stack

Apart from that, I think this is a more correct approach than #598, especially accounting for the new keyword definition format in DSL v2.

The main change is that we only check the keyword trie and additionally the (newly introduced) compound keyword scanners only after the token has been lexed as an identifier. For each context, we collect Identifier scanners used by the keywords and attempt promotion there.

The existing lexing performance is not impacted from what I've seen when running the sanctuary tests and I can verify (incl. CST tests) that we now properly parse source that uses contextual keywords (e.g. `from`) and that the compound keywords (e.g. `ufixedMxN`) are properly versioned.

This adapts the existing `codegen_grammar` interface that's a leftover from DSLv1; I did that to work on finishing #638; once this is merged and we now properly parse contextual keywords, I'll move to clean it up and reduce the parser codegen indirection (right now we go from v2 -> v1 model -> code generator -> Tera templates; it'd like to at least cut out the v1 model and/or simplify visiting v2 from the existing `CodeGenerator`).

Please excuse the WIP comments in the middle; the first and the last ones should make sense when reviewing. I can simplify this a bit for review, if needed.

### CODE REVIEW

File: crates/codegen/parser/generator/src/code_generator.rs [hunk 1]
```diff
@@ -28,27 +28,31 @@ pub struct CodeGenerator {
...
     scanner_definitions: BTreeSet<&'static str>,
-    alpha_literal_scanner: String,
-    non_alpha_literal_scanner: String,
+    literal_scanner: String,
+    keyword_compound_scanners: BTreeMap<&'static str, String>,
+    keyword_trie_scanner: String,
+    #[serde(skip)]
+    keyword_scanner_defs: BTreeMap<&'static str, KeywordScannerDefinitionRef>,
+    /// Name of the scanners that can be promoted to a keyword
+    identifier_scanners: BTreeSet<&'static str>,
```
Comment (AntonyBlakey) [inline crates/codegen/parser/generator/src/code_generator.rs:55]: If a name needs a comment to make up for the lack of sufficient naming, then something is wrong IMO. I believe that a long descriptive name is better than a short name that needs a comment.

**Review by AntonyBlakey — CHANGES_REQUESTED** [review]

File: crates/codegen/parser/generator/src/code_generator.rs (hunk 1 above)
Comment (Xanewok) [reply crates/codegen/parser/generator/src/code_generator.rs:55]: In a general context, I agree.

Here, however, the comment tries to enrichen the context, as keywords are already inherently "promotable" from the identifiers (no new information); this is just a reminder to the reader and hints how it's used later on in the template expansion. I'll remove it if you think it's redundant (the alternative of `identifier_scanner_names_promotable_to_keywords` seems a bit too extra and redundant, as per above).

Comment (Xanewok) [reply crates/codegen/parser/generator/src/code_generator.rs]: I've renamed from `identifier_scanners` to `identifier_scanner_names` in https://github.com/NomicFoundation/slang/pull/723/commits/581d611f291c64a7a9f7ee3133362ad105c78e67; this is more symmetrical with other `*_scanner_names` and carries some "new" context over from the comment above; hopefully that's a good middle ground!

**Comment by Xanewok** [comment]
Instead of `next_token` returning `Vec` (first pass), we now introduce the following type:
```rust
pub enum ScannedToken {
    Single(TokenKind),
    IdentifierOrKeyword {
        identifier: TokenKind,
        kw: KeywordScan,
    },
}
```

Not only we don't go through heap for the often-called `next_token`, we now explicitly have to handle the case of identifiers/keywords and whether a keyword is strictly reserved or "present", so usable in both keyword and identifier position.

To keep the existing parser flow mostly the same, a helper fn `ScannedToken::unambiguous(self) -> TokenKind` is introduced, that tries to map back to a single token kind if possible (for the "present" keyword it falls back to a general "identifier" token kind).

This also introduces another helper fn `ScannedToken::accepted_as(self, expected: TokenKind) -> bool` that is used in the parser functions as a convenience, which returns whether the scanned token is accepted in a given position, streamlining the underlying additional kw reservation checks for identifier token kinds.

File: crates/codegen/parser/runtime/src/support/scanner_macros.rs [hunk 2]
```diff
@@ -66,6 +66,26 @@ macro_rules! scan_choice {
...
 
+#[allow(unused_macros)]
+macro_rules! scan_keyword_choice {
+    ($stream:ident, $ident:ident, $($scanner:expr),*) => {
+        loop {
+            let save = $stream.position();
+            $(
+                {
+                    if let result @ (KeywordScan::Present(..) | KeywordScan::Reserved(..)) = ($scanner) {
+                        if $ident.len() == $stream.position().utf8 - save.utf8 {
```
Comment (Xanewok) [inline crates/codegen/parser/runtime/src/support/scanner_macros.rs:77]: Not perfect, I admit; this should probably be handled with an automaton for the keywords and executed/matched on the string till it's exhausted but let's punt on this for now

File: crates/codegen/parser/runtime/src/templates/language.rs.jinja2 [hunk 3]
```diff
@@ -219,11 +267,6 @@ impl Language {
         return Self::SUPPORTED_VERSIONS.iter().map(|v| v.to_string()).collect();
     }
 
-    #[napi(js_name = "scan", ts_return_type = "kinds.TokenKind | null", catch_unwind)]
```
Comment (Xanewok) [inline crates/codegen/parser/runtime/src/templates/language.rs.jinja2:222]: I should probably separate a PR that gets rid of the `scan` API, as outlined in #640

File: crates/codegen/parser/runtime/src/templates/mod.rs.jinja2 [hunk 4]
```diff
@@ -5,7 +5,7 @@ pub mod cst;
 pub mod cursor;
 pub mod kinds;
 pub mod language;
-pub(crate) mod lexer;
+pub mod lexer;
```
Comment (Xanewok) [inline crates/codegen/parser/runtime/src/templates/mod.rs.jinja2:8]: We need to expose `ScannedToken` as it's a part of the public API now; we should either hide the lexer-related API and tests as unit tests (rather than integration ones) or just expose this module anyway

Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/templates/mod.rs.jinja2:8]: > we should either hide the lexer-related API

Is that the part of #640 about hiding the `scan()` API? I suggest doing so, so that we don't have to expose `ScannedToken` or this machinery at all. It is a rather internal solution that will change (frequently) over time, and I don't think there is any (reported) user benefit for it right now. I think we will need to think of how we can expose this long term once we have embedded language support.

Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/templates/mod.rs.jinja2:8]: > tests as unit tests (rather than integration ones)

I think we have very few lexer-related tests. We can either move them to snapshot tests (testing via a calling parser), or if not possible for some corner case, move it to unit tests like you suggested.

File: crates/codegen/parser/runtime/src/templates/language.rs.jinja2 (hunk 3 above)
Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/templates/language.rs.jinja2:222]: +1 .. if that means we no longer have to expose `ScannedToken` or similar APIs.

File: crates/solidity/outputs/npm/package/src/generated/index.d.ts [hunk 5]
```diff
@@ -519,7 +519,6 @@ export namespace language {
     constructor(version: string);
     get version(): string;
     static supportedVersions(): Array<string>;
-    scan(lexicalContext: LexicalContext, input: string): kinds.TokenKind | null;
     parse(kind: kinds.RuleKind, input: string): parse_output.ParseOutput;
```
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/package/src/generated/index.d.ts:522]: I suggest adding a changeset about the removed API.

File: crates/solidity/testing/snapshots/cst_output/EventDefinition/transfer/generated/0.4.11-success.yml [hunk 6]
```diff
@@ -0,0 +1,37 @@
+# This file is generated automatically by infrastructure scripts. Please don't edit by hand.
+
+Source: >
+  1  │ event Transfer(address indexed from, address indexed to, uint256 value);         │ 0..72
```
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/EventDefinition/transfer/generated/0.4.11-success.yml:4]: Question about error recovery, unrelated to this PR:

In the before (deleted) snapshot, shouldn't the `SeparetedBy` error recovery have recovered from the error on `from` here, to skip it, and recover after the following comma, to continue parsing the next parameter `address indexed to` correctly? Or is this NIY?

File: crates/solidity/testing/snapshots/cst_output/YulVariableDeclarationStatement/keyword_ufixed8x8/input.sol [hunk 7]
```diff
@@ -0,0 +1 @@
+let ufixed8x8
```
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/YulVariableDeclarationStatement/keyword_ufixed8x8/input.sol:1]: nit suggestion: adding a comment here (and similar inputs) to describe the specific corner case we are testing, to make it easier to spot mistakes in future PRs if something was accidentally changed.

```suggestion
// 'ufixed8x8' should be reserved till 0.7.1:
let ufixed8x8
```

**Review by OmarTawfik — COMMENTED** [review]
Left a few comments.

File: crates/codegen/parser/runtime/src/templates/language.rs.jinja2 (hunk 3 above)
Comment (Xanewok) [reply crates/codegen/parser/runtime/src/templates/language.rs.jinja2:222]: Separated this as https://github.com/NomicFoundation/slang/pull/728

File: crates/codegen/parser/runtime/src/templates/mod.rs.jinja2 (hunk 4 above)
Comment (Xanewok) [reply crates/codegen/parser/runtime/src/templates/mod.rs.jinja2:8]: Separated in https://github.com/NomicFoundation/slang/pull/728; I'll revert the changes unrelated changes here. The lexer test has to be a unit test, since we directly want to test the immediate behaviour of `next_token`

File: crates/solidity/outputs/npm/package/src/generated/index.d.ts (hunk 5 above)
Comment (Xanewok) [reply crates/solidity/outputs/npm/package/src/generated/index.d.ts:522]: Done in https://github.com/NomicFoundation/slang/pull/728

File: crates/codegen/parser/generator/src/code_generator.rs (hunk 1 above)
Comment (Xanewok) [reply crates/codegen/parser/generator/src/code_generator.rs]: Done in https://github.com/NomicFoundation/slang/pull/723/commits/2336ad9538a04dcffaace368953272212bfa1a96

File: crates/solidity/testing/snapshots/cst_output/YulVariableDeclarationStatement/keyword_ufixed8x8/input.sol (hunk 7 above)
Comment (Xanewok) [reply crates/solidity/testing/snapshots/cst_output/YulVariableDeclarationStatement/keyword_ufixed8x8/input.sol:2]: Done in https://github.com/NomicFoundation/slang/pull/723/commits/322f105dca48abc08f844e6fdfde296d169d17b6

File: crates/solidity/testing/snapshots/cst_output/EventDefinition/transfer/generated/0.4.11-success.yml (hunk 6 above)
Comment (Xanewok) [reply crates/solidity/testing/snapshots/cst_output/EventDefinition/transfer/generated/0.4.11-success.yml:4]: If I recall correctly/not mixing things up, we couldn't properly recover from `NoMatch` elements in a `SeparatedBy` parser without risking misparses, unless we did the FIRST/FOLLOW computation that we decided to punt on for now. I had a counter-example but I don't have it handy right now...

Similar behaviour can be noticed here: https://github.com/Xanewok/slang/blob/8fe0b94218db07163f33e09cc4d15288e7f48b42/crates/solidity/testing/snapshots/cst_output/ContractMembers/separated_recovery/generated/0.4.11-failure.yml#L86

However, partial `IncompleteMatch` are recovered from until the separator, see https://github.com/Xanewok/slang/blob/8fe0b94218db07163f33e09cc4d15288e7f48b42/crates/solidity/testing/snapshots/cst_output/ContractMembers/separated_recovery/generated/0.4.11-failure.yml#L78

Then again, I would imagine a sequence of (TypeName, IndexedKeyword, Identifier) should be marked as "incomplete" (not empty/no match) if missing the suffix, so I may be misremembering the issue there. Definitely worth looking into it!

Comment (OmarTawfik) [reply crates/solidity/testing/snapshots/cst_output/EventDefinition/transfer/generated/0.4.11-success.yml:4]: Thanks for explaining!

**Review by AntonyBlakey — APPROVED** [review]

**Review by OmarTawfik — APPROVED** [review]
